File: php/22-generate-parentheses.php
<?php
class Solution {
    private $res = [];
    /**
     * @param Integer $n
     * @return String[]
     */
    function generateParenthesis($n) {
        $this->dfs($n, 0, 0);
        return $this->res;
    }
    function dfs($n, $left, $right, $p = '') {
        if ($left >= $n && $right >= $n) {
            $this->res[] = $p;
            return;
        }
        if ($left < $n) {
            $this->dfs($n, $left + 1, $right, $p . '(');
        }
        if ($right < $left) {
            $this->dfs($n, $left, $right + 1, $p . ')');
        }
    }
}

File: go/22-generate-parentheses.go
package main

func generateParenthesis(n int) []string {
	res := make([]string, 0)
	dfs(n, 0, 0, "", &res)
	return res
}
func dfs(n, left, right int, p string, res *[]string) {
	if left >= n && right >= n {
		*res = append(*res, p)
		return
	}
	if left < n {
		dfs(n, left+1, right, p+"(", res)
	}
	if right < left {
		dfs(n, left, right+1, p+")", res)
	}
}
